src/builder/commands/copy.rs

```diff
@@ -159,11 +185,33 @@ impl BuildStep for Copy {
     }
 }
 
-fn hash_path(hash: &mut Digest, path: &Path,
-    filter: &Filter, owner_uid: Option<uid_t>, owner_gid: Option<gid_t>)
+fn reset_permissions(path: &Path, umask: u32)
+    -> Result<(), StepError>
+{
+    let stat = try!(path.symlink_metadata()
+        .map_err(|e| StepError::Read(PathBuf::from(path), e)));
+    if stat.is_dir() {
+        let perms = Permissions::from_mode(DIR_MODE & !umask);
+        try!(set_permissions(path, perms)
+            .map_err(|e| StepError::Write(PathBuf::from(path), e)));
+    } else if stat.is_file() {
+        let orig_mode = stat.permissions().mode();
+        let mode = if orig_mode & EXE_CHECK_MASK > 0 {
+            DIR_MODE
```

Comment:
DIR_MODE? It would be more readable if you put `0o777` here, rather than a constant with a misleading name